### PR TITLE
Initial multiplatform migration steps

### PR DESCRIPTION
**Pull request by LaatonWalaBhoot**
- Migrated to Ktor-Okhttp 2.3.4
- Migrated to Ksoup 0.0.2

### CODE REVIEW

File: unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt [hunk 1]
```diff
@@ -33,8 +34,8 @@ open class HtmlTagsBasedUnfurler : UnfurlerExtension {
 
         if (body != null && body.contentType().isHtmlText()) {
           // TODO: stream the HTML body only until a "</head>" is received instead of streaming the entire HTML body.
-          Jsoup.parse(
-            /* in */ body.source().inputStream(),
+          Ksoup.parse(
+            /* in */ BufferReader(body.bytes()),
```
Comment (saket) [inline unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: Does `ksoup` have a streaming API? Calling `bytes()` is dangerous here as it'll allocate the entire HTML in memory.

File: unfurl/build.gradle.kts [hunk 2]
```diff
@@ -5,8 +5,8 @@ plugins {
 apply(plugin = "com.vanniktech.maven.publish")
 
 dependencies {
-  api("com.squareup.okhttp3:okhttp:4.9.0")  // Updating to 5.x? See cli/build.gradle.kts.
-  api("org.jsoup:jsoup:1.16.1")
+  api("io.ktor:ktor-client-okhttp:2.3.4")
+  api("com.fleeksoft.ksoup:ksoup:0.0.2")
```
Comment (saket) [inline unfurl/build.gradle.kts]: Have you compared this with https://github.com/MohamedRejeb/Ksoup?

**Review by saket — CHANGES_REQUESTED** [review]

File: unfurl/build.gradle.kts (hunk 2 above)
Comment (LaatonWalaBhoot) [reply unfurl/build.gradle.kts]: Yes, this one does not have support for Element and Document, only Parsing hence would not be able to fulfill our purpose

File: unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt (hunk 1 above)
Comment (LaatonWalaBhoot) [reply unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: Have switched this to buffer. Can you take a look?

File: unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt [hunk 3]
```diff
@@ -33,8 +34,8 @@ open class HtmlTagsBasedUnfurler : UnfurlerExtension {
 
         if (body != null && body.contentType().isHtmlText()) {
           // TODO: stream the HTML body only until a "</head>" is received instead of streaming the entire HTML body.
-          Jsoup.parse(
-            /* in */ body.source().inputStream(),
+          Ksoup.parse(
+            /* in */ BufferReader(body.source().buffer),
```
Comment (saket) [inline unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: Bad news: `ksoup` doesn't actually implement streaming parsing. It internally allocates the entire content in memory.

https://github.com/fleeksoft/ksoup/blob/7365999c33d9f9ba37d9c3bde64fc63f7eff561a/ksoup/src/commonMain/kotlin/com/fleeksoft/ksoup/ported/BufferReader.kt#L35

We'll unfortunately not be able to use `ksoup` until they fix this problem 😕

Comment (saket) [reply unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: There may be another option: `unfurl` only needs the `<head>` portion of web pages. Can `unfurl` stream the HTML until a `</head>` is found and then give it to `ksoup`? It'll still allocate everything in memory but at least the allocation will be significantly smaller.

Comment (saket) [reply unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: I filed an issue here: https://github.com/fleeksoft/ksoup/issues/2

Comment (LaatonWalaBhoot) [reply unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: Saw that...Lets wait for him to reply.
Handling streaming in the library itself could work, however ideally would want the parser to handle this logic

Comment (LaatonWalaBhoot) [reply unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: @saket 
Can we make other changes in the meantime?
Such as overhauling the library to a multiplatform one and making a dev release?
That way we will be sure everything else is in place

Comment (saket) [reply unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: Will it only contain JVM sources then?

Comment (LaatonWalaBhoot) [reply unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: No, I meant. 
We will continue with ksoup only
and make other changes on a separate branch such as Ktor and source sets
Wait for ksoup to come up with a change, else I will try to submit a PR for them

Comment (saket) [reply unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: Unfortunately we can't move forward anymore until `ksoup` fixes this. Merging `ksoup` would prevent me from making any releases to `unfurl` until they fix their streaming API.

Comment (LaatonWalaBhoot) [reply unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: Oh, I mean to continue without merging
Keep this PR updated in any case.
And merge only when Ksoup is on par.
That way you can make newer releases and we can make progress on this as well

Comment (saket) [reply unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: Long lived branches aren't great so I'd not suggest that. Wanna migrate to ktor only instead?

Comment (LaatonWalaBhoot) [reply unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: Yes. Lets do this then.
Should I revert Ksoup changes for now?
And use Ktor CIO?

Comment (LaatonWalaBhoot) [reply unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: @saket I have updated to Ksoup 0.0.5 with streaming support
Can you take a look?

Comment (LaatonWalaBhoot) [reply unfurl/src/main/kotlin/me/saket/unfurl/extension/HtmlTagsBasedUnfurler.kt]: @saket Have upgraded to 0.0.6.
Ksoup has migrated to input stream for HTML 
Can you take a look?

**Comment by saket** [comment]
Any reason for closing this already?

**Comment by LaatonWalaBhoot** [comment]
Please follow this here: https://github.com/saket/unfurl/pull/3
made further changes.
Migrating the Android library to KMP in the same branch was a huge headache.
Hence created a new project to properly migrate and then evened out the branches.

**Pull request closed without merging**